File: lua/ht/conf/lsp/servers/flutter.lua
local RightClick = require("ht.core.right-click")
local CoreLspServer = require("ht.core.lsp.server")

local function t_cmd(s, cmd, keys)
  return {
    s,
    callback = function()
      vim.cmd(cmd)
    end,
    keys = keys,
  }
end

local function f_cmd(title, cmd)
  return {
    title = title,
    f = function()
      vim.cmd(cmd)
    end,
  }
end

---@type ht.lsp.ServerOpts
local opts = {}

opts.name = "flutter"

opts.mason = false

opts.setup = function(on_buffer_attach, capabilities)
  require("flutter-tools").setup {
    ui = {
      border = "solid",
      notification_style = "plugin",
    },
    debugger = {
      enabled = false,
    },
    widget_guides = {
      enabled = false,
    },
    dev_log = {
      enabled = true,
    },
    lsp = {
      color = {
        enabled = true,
      },
      on_attach = on_buffer_attach,
      capabilities = capabilities,
      settings = {
        showTodos = false,
        enableSnippets = true,
        completeFunctionCalls = true,
      },
    },
  }
  require("telescope").load_extension("flutter")
end

opts.right_click = {
  {
    index = RightClick.indexes.flutter,
    enabled = {
      filetype = {
        "dart",
      },
    },
    items = {
      t_cmd("Flutter: Run", "FlutterRun"),
      t_cmd("Flutter: Hot Restart", "FlutterRestart", { "R" }),
      {
        "Flutter More",
        keys = { "F" },
        children = {
          t_cmd("Hot Reload", "FlutterReload"),
          t_cmd("Toggle Outline", "FlutterOutlineToggle"),
          t_cmd("Open Outline", "FlutterOutlineOpen"),
          t_cmd("Go to super class", "FlutterSuper"),
        },
      },
    },
  },
}

opts.function_sets = {
  {
    category = "Flutter",
    ---@param buffer VimBuffer
    filter = function(buffer)
      for _, server in ipairs(buffer.lsp_servers) do
        if server.name == "dartls" then
          return true
        end
      end
      return false
    end,
    functions = {
      f_cmd(
        "Run the current project. This needs to be run from within a flutter project",
        "FlutterRun"
      ),
      f_cmd(
        "Brings up a list of connected devices to select from",
        "FlutterDevices"
      ),
      f_cmd(
        "Similar to devices but shows a list of emulators to choose from",
        "FlutterEmulators"
      ),
      f_cmd("Reload the running project", "FlutterReload"),
      f_cmd("Restart the current project", "FlutterRestart"),
      f_cmd("Ends a running session", "FlutterQuit"),
      f_cmd(
        "Ends a running session locally but keeps the process running on the device",
        "FlutterDetach"
      ),
      f_cmd(
        "Toggle the outline window showing the widget tree for the given file",
        "FlutterOutlineToggle"
      ),
      f_cmd(
        "Opens an outline window showing the widget tree for the given file",
        "FlutterOutlineOpen"
      ),
      f_cmd("Starts a Dart Dev Tools server", "FlutterDevTools"),
      f_cmd("Activates a Dart Dev Tools server", "FlutterDevToolsActivate"),
      f_cmd(
        "Copies the profiler url to your system clipboard (+ register). Note that commands FlutterRun and FlutterDevTools must be executed first",
        "FlutterCopyProfilerUrl"
      ),
      f_cmd(
        "This command restarts the dart language server, and is intended for situations where it begins to work incorrectly",
        "FlutterLspRestart"
      ),
      f_cmd(
        "Go to super class, method using custom LSP method dart/textDocument/super",
        "FlutterSuper"
      ),
      f_cmd(
        "Forces LSP server reanalyze using custom LSP method dart/reanalyze",
        "FlutterReanalyze"
      ),
      f_cmd(
        [[Renames and updates imports if lsp.settings.renameFilesWithClasses == "always"]],
        "FlutterRename"
      ),
    },
  },
}

return CoreLspServer.new(opts)

File: src/conf/external_tools/lsp_servers/flutter.ts
import {
  ActionGroupBuilder,
  LspServer,
  Plugin,
  PluginOptsBase,
  andActions,
} from "@core/model";

const spec: PluginOptsBase = {
  shortUrl: "akinsho/flutter-tools.nvim",
  lazy: {
    lazy: true,
    dependencies: [
      "nvim-lua/plenary.nvim",
      "stevearc/dressing.nvim",
      "nvim-telescope/telescope.nvim",
    ],
  },
};

function generateActions() {
  return ActionGroupBuilder.start()
    .from("flutter-tools.nvim")
    .category("Flutter")
    .condition((buffer) => {
      return buffer.filetype === "dart";
    })
    .addOpts({
      id: "flutter-tools.run-project",
      title:
        "Run the current project. This need to be run from within" +
        " a flutter project",
      callback: "FlutterRun",
    })
    .addOpts({
      id: "flutter-tools.select-devices",
      title: "Brings up a list of connected devices to select from",
      callback: "FlutterDevices",
    })
    .addOpts({
      id: "flutter-tools.select-emulators",
      title: "Similar to devices but shows a list of emulators to choose from",
      callback: "FlutterEmulators",
    })
    .addOpts({
      id: "flutter-tools.hot-reload",
      title: "Reload the running project",
      callback: "FlutterReload",
    })
    .addOpts({
      id: "flutter-tools.restart-project",
      title: "Restart the current project",
      callback: "FlutterRestart",
    })
    .addOpts({
      id: "flutter-tools.quit-project",
      title: "Ends a running session",
      callback: "FlutterQuit",
    })
    .addOpts({
      id: "flutter-tools.detach-project",
      title:
        "Ends a running session, but keeps the process running on the device",
      callback: "FlutterDetach",
    })
    .addOpts({
      id: "flutter-tools.toggle-outline",
      title:
        "Toggle the outline window showing the widget tree for the given file",
      callback: "FlutterOutlineToggle",
    })
    .addOpts({
      id: "flutter-tools.open-outline",
      title:
        "Opens an outline window showing the widget tree for the given file",
      callback: "FlutterOutlineOpen",
    })
    .addOpts({
      id: "flutter-tools.start-dev-tools",
      title: "Starts a Dart Dev Tools server",
      callback: "FlutterDevTools",
    })
    .addOpts({
      id: "flutter-tools.active-dev-tools",
      title: "Activates a Dart Dev Tools server",
      callback: "FlutterDevToolsActivate",
    })
    .addOpts({
      id: "flutter-tools.copy-profiler-url",
      title:
        "Copies the profiler url to your system clipboard (+ register). Note that commands FlutterRun and FlutterDevTools must be executed first",
      callback: "FlutterCopyProfilerUrl",
    })
    .addOpts({
      id: "flutter-tools.lsp-restart",
      title:
        "This command restarts the dart language server, and is intended for situations where it begins to work incorrectly",
      callback: "FlutterLspRestart",
    })
    .addOpts({
      id: "flutter-tools.goto-super-class",
      title:
        "Go to super class, method using custom LSP method dart/textDocument/super",
      callback: "FlutterSuper",
    })
    .addOpts({
      id: "flutter-tools.reanalyze",
      title:
        "Forces LSP server reanalyze using custom LSP method dart/reanalyze",
      callback: "FlutterReanalyze",
    })
    .addOpts({
      id: "flutter-tools.rename",
      title:
        "Renames and updates imports if lsp.settings.renameFilesWithClasses == 'always'",
      callback: "FlutterRename",
    })
    .build();
}

const plugin = new Plugin(andActions(spec, generateActions));

export const server = new LspServer({
  name: "flutter",
  plugin,
  exe: false,
  setup: (_, on_attach, capabilities) => {
    luaRequire("flutter-tools").setup({
      ui: {
        border: "solid",
        notification_style: "plugin",
      },
      debugger: {
        enabled: false,
      },
      widget_guides: {
        enabled: false,
      },
      dev_log: {
        enabled: true,
      },
      lsp: {
        color: {
          enabled: true,
        },
        on_attach: on_attach,
        capabilities: capabilities,
        settings: {
          showTodos: false,
          enableSnippets: true,
          completeFunctionCalls: true,
        },
      },
    });
    luaRequire("telescope").load_extension("flutter");
  },
});
